README.md
cabag_loginas
=============

Friendly fork of current TER version without support for TYPO3 CMS 6.2+ and with improved styling.

* Within the backend you have a button in the fe_user table to quickly login as this fe user in frontend.
* If you have a frontend user with the same email adress as your backend user you get an icon in the top toolbar menu.
* If you have several fe_user accounts with the same email address you get the all shown.
* You can configure the information displayed with the icon in the EM.

You may provide pull requests if you think that anything can be improved.

ChangeLog
2018-04-25 Tizian Schmidlin <st@cabag.ch>
	* version 2.1.3
	* fixed issue with php7 strict in LoginAsService
	* remove sorting from ToolbarItemHook
	* Thanks for Uwe Michelfelder for his quick feedback

2018-04-19 Tizian Schmidlin <st@cabag.ch>
	* version 2.1.2
	* check if user id is set before returning a href

2018-01-29 Christopher Schnell <cs@cabag.ch>
	* version 2.1.1 of cabag_loginas
    * Fixed Bug in composer.json

2018-01-23 Christopher Schnell <cs@cabag.ch>
	* version 2.1.0 of cabag_loginas
	* Starting to move towards T3V9
    * Moved TCA Overrides to their correct folders
    * removed old comments and cleaned up
    * dropped 6.2 compatibility

2017-11-27 Xiaoling Peng <xp@cabag.ch>
	* version 2.0.5
	* fixed Classes/Hook/RecordListHook.php: makeControl to remove foreach ($cells as $key => $value) - #43266

2017-11-23 Surena Golijani <sg@cabag.ch>
	* version 2.0.4
	* edited Classes/Hook/ToolbarItemHook.php getLink(): check if form of an unsaved new user is displayed to prevent error: Unknown column 'NEW5a168f878d66b716571009' in 'where clause', working on issue #43251

2016-07-28 Tizian Schmidlin <st@cabag.ch>
	* version 2.0.3
	* fixed tooltip
	* version 2.0.2
	* made the backend icon to switch user look fancy

2016-07-26 Tizian Schmidlin <st@cabag.ch>
	* version 2.0.1
	* fixed all issues with 7.6
	* The icon is still ugly, though

2016-05-04 Tizian Schmidlin <st@cabag.ch>
	* version 2.0.0
	* merged https://github.com/visol/ext-cabag_loginas and https://github.com/spoonerWeb/ext-cabag_loginas

2014-09-30 Nils Blattner <nb@cabag.ch>
	* version 1.2.0
	* [6.2] fixed compatibility issues with TYPO3 6.2
	* [6.2] set TYPO3 dependency to 4.5.0-6.2.99
	* [TASK] removed trailing ?> from php files

2013-07-24 Nicole Cordes <typo3@cordes.co>
	* version 1.1.9
	* fixed respecting of domain records for multi domain systems

2013-07-23 Nicole Cordes <typo3@cordes.co>
	* version 1.1.8
	* format code according to coding guidelines

2013-04-09 Toby Merz <tm@cabag.ch>
	* version 1.1.7
	* adapted extension to clean up requirements for TER

2012-02-15 Dimitri Koenig  <dk@cabag.ch>
	* version 1.1.6
	* added loginas field in fe_users table due to TYPO3 4.5 tca conventions

2011-10-06 Nils Blattner <nb@cabag.ch>
	* version 1.1.5
	* fixed a bug when no domain record based redirect was set

2011-09-21 Nils Blattner <nb@cabag.ch>
	* version 1.1.4
	* added the possibility to redirect after the login to a specific page, based on the domain
	* added an option in ext_conf_template.txt to enable the additional fields in sys_domain (enableDomainBasedRedirect)
	* added the new field to ext_tables.sql and ext_tables.php

2011-06-29 Dimitri Koenig  <dk@cabag.ch>
	* version 1.1.3
	* fixed bug in hashkey generation

2011-05-11 Nils Blattner <nb@cabag.ch>
	* version 1.1.2
	* changed sv1 to store a value in the session data, so that loginas users could be tracked

2011-02-04 Nils Blattner <nb@cabag.ch>
	* version 1.1.1
	* fixed cabag_loginas.js so it really does not produce a JS-Error when the loginas-menu is not present

2010-11-16 Nils Blattner <nb@cabag.ch>
	* version 1.1.0
	* fixed cabag_loginas.js so it does not produce a JS-Error when the loginas-menu is not present

2010-09-07 Dimitri Koenig  <dk@cabag.ch>
	* version 1.0.10
	* limited display of FE users to 15, just in case you have over 400.000 fe_users with the same email address. Thx to Nils Blattner :-)

2010-07-02 Dimitri Koenig  <dk@cabag.ch>
	* version 1.0.9
	* changed fixed http:// path to relative path

2010-07-02 Dimitri Koenig  <dk@cabag.ch>
	* version 1.0.8
	* ter upload
	* extended timeout from 300 to 3600 seconds

2010-03-19 Dimitri Koenig  <dk@cabag.ch>
	* version 0.0.7
	* better toolbar menu allowing multiple fe_user accounts with the same email address to be shown

2010-03-16 Dimitri Koenig  <dk@cabag.ch>
	* version 0.0.6
	* small css change

2010-03-16 Dimitri Koenig  <dk@cabag.ch>
	* version 0.0.5
	* additional security
	* toolbar button if a fe_user exist with same email address as backend user

2010-02-15 Dimitri Koenig  <dk@cabag.ch>
	* version 0.0.4
	* username will be shown instead of name

2010-02-15 Dimitri Koenig  <dk@cabag.ch>
	* version 0.0.3
	* fixed display icons bug

2010-02-15 Dimitri Koenig  <dk@cabag.ch>
	* version 0.0.2
	* fixed auth service bug

2010-02-15 Dimitri Koenig  <dk@cabag.ch>
	* version 0.0.1
	* initial code generated with kickstarter


class.tx_cabagloginas.6.2.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2010 Dimitri König <dk@cabag.ch>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *  A copy is found in the textfile GPL.txt and important notices to the license
 *  from the author is found in LICENSE.txt distributed with these scripts.
 *
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

class tx_cabagloginas implements \TYPO3\CMS\Backend\Toolbar\ToolbarItemHookInterface {
	protected $backendReference;

	protected $users = array();

	protected $EXTKEY = 'cabag_loginas';

	public function __construct(\TYPO3\CMS\Backend\Controller\BackendController &$backendReference = NULL) {
		$GLOBALS['LANG']->includeLLFile('EXT:cabag_loginas/locallang_db.xml');
		$this->backendReference = $backendReference;

		$email = $GLOBALS['BE_USER']->user['email'];

		$this->users = $GLOBALS['TYPO3_DB']->exec_SELECTgetRows(
			'*',
			'fe_users',
			'email = ' . $GLOBALS['TYPO3_DB']->fullQuoteStr($email, 'fe_users') . ' AND disable = 0 AND deleted = 0',
			'',
			'',
			'15'
		);
	}

	public function checkAccess() {
		$conf = $GLOBALS['BE_USER']->getTSConfig('backendToolbarItem.tx_cabagloginas.disabled');

		return ($conf['value'] == 1 ? FALSE : TRUE);
	}

	public function render() {
		$this->backendReference->addCssFile('cabag_loginas', t3lib_extMgm::extRelPath($this->EXTKEY) . 'cabag_loginas.css');
		$this->backendReference->addJavascriptFile(t3lib_extMgm::extRelPath($this->EXTKEY) . 'cabag_loginas.js');

		$toolbarMenu = array();

		$title = $GLOBALS['LANG']->getLL('fe_users.tx_cabagloginas_loginas', TRUE);
		$ext_conf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['cabag_loginas']);
		$defLinkText = trim($ext_conf['defLinkText']);
		if (empty($defLinkText) || strstr($defLinkText, '#') === FALSE || strstr($defLinkText, 'password') !== FALSE) {
			$defLinkText = '[#pid# / #uid#] #username# (#email#)';
		}

		if (count($this->users)) {
			if (count($this->users) == 1) {
				$title .= ' ' . $this->formatLinkText($this->users[0], $defLinkText);
				$toolbarMenu[] = $this->getLoginAsIconInTable($this->users[0], $title);
			} else {
				$toolbarMenu[] = '<a href="#" class="toolbar-item"><img' . t3lib_iconWorks::skinImg($this->backPath, 'gfx/su_back.gif', 'width="16" height="16"') . ' title="' . $title . '" alt="' . $title . '" /></a>';

				$toolbarMenu[] = '<ul class="toolbar-item-menu" style="display: none;">';

				foreach ($this->users as $user) {
					$linktext = $this->formatLinkText($user, $defLinkText);
					$link = $this->getHREF($user);
					$toolbarMenu[] = '<li><a href="' . htmlspecialchars($link) . '" target="_blank"><img' . t3lib_iconWorks::skinImg($this->backPath, 'gfx/i/fe_users.gif', 'width="16" height="16"') . ' title="' . $title . '" alt="' . $title . '" /> ' . $linktext . '</a></li>';
				}

				$toolbarMenu[] = '</ul>';
			}

			return implode("\n", $toolbarMenu);
		}
	}

	public function formatLinkText($user, $defLinkText) {
		foreach ($user as $key => $value) {
			$defLinkText = str_replace('#' . $key . '#', $value, $defLinkText);
		}

		return $defLinkText;
	}

	public function getAdditionalAttributes() {
		if (count($this->users)) {
			return ' id="tx-cabagloginas-menu"';
		} else {
			return '';
		}
	}

	function getHREF($user) {
		$parameterArray = array();
		$parameterArray['userid'] = (string) $user['uid'];
		$parameterArray['timeout'] = (string) $timeout = time() + 3600;
		// Check user settings for any redirect page
		// Added "" for user uid
		if ($user['felogin_redirectPid']) {
			$parameterArray['redirecturl'] = $this->getRedirectUrl($user['felogin_redirectPid']);
		} else {
			// Check group settings for any redirect page
			$userGroup = $GLOBALS['TYPO3_DB']->exec_SELECTgetSingleRow(
				'fe_groups.felogin_redirectPid',
				'fe_users, fe_groups',
				'fe_groups.felogin_redirectPid != "" AND fe_groups.uid IN (fe_users.usergroup) AND fe_users.uid = "' . $user['uid'].'"'
			);
			if (is_array($userGroup) && !empty($userGroup['felogin_redirectPid'])) {
				$parameterArray['redirecturl'] = $this->getRedirectUrl($userGroup['felogin_redirectPid']);
			} elseif (rtrim(t3lib_div::getIndpEnv('TYPO3_SITE_URL'), '/') !== ($domain = $this->getRedirectForCurrentDomain($user['pid']))) {
				// Any manual redirection defined in sys_domain record
				$parameterArray['redirecturl'] = rawurlencode($domain);
			}
		}
		$ses_id = $GLOBALS['BE_USER']->user['ses_id'];
		$parameterArray['verification'] = md5($GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey'] . $ses_id . serialize($parameterArray));
		$link = t3lib_div::getIndpEnv('TYPO3_SITE_URL') . '?' . t3lib_div::implodeArrayForUrl('tx_cabagloginas', $parameterArray);

		return $link;
	}

	function getLink($data) {
		$label = $data['label'] . ' ' . $data['row']['username'];
		$link = $this->getHREF($data['row']);
		$content = '<a href="' . $link . '" target="_blank" style="text-decoration:underline;">' . $label . '</a>';

		return $content;
	}

	function getLoginAsIconInTable($user, $title = '') {
		$label = '<img' . t3lib_iconWorks::skinImg($this->backPath, 'gfx/su_back.gif', 'width="16" height="16"') . ' title="' . $title . '" alt="' . $title . '" />';
		$link = $this->getHREF($user);
		$content = '<a class="toolbar-item" href="' . $link . '" target="_blank">' . $label . '</a>';

		return $content;
	}

	/**
	 * Finds the redirect link for the current domain.
	 *
	 * @param integer $pid Page id the user is stored in
	 *
	 * @return string '../' if nothing was found, the link in the form of http://www.domain.tld/link/page.html otherwise.
	 */
	function getRedirectForCurrentDomain($pid) {
		$extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['cabag_loginas']);
		$domain = t3lib_BEfunc::getViewDomain($pid);
		$domainArray = parse_url($domain);

		if (empty($extConf['enableDomainBasedRedirect'])) {
			return $domain;
		}

		$rowArray = $GLOBALS['TYPO3_DB']->exec_SELECTgetRows(
			'domainName, tx_cabagfileexplorer_redirect_to',
			'sys_domain',
			'hidden = 0 AND domainName = ' . $GLOBALS['TYPO3_DB']->fullQuoteStr($domainArray['host'], 'sys_domain'),
			'',
			'',
			1
		);

		if (count($rowArray) === 0 || (trim($rowArray[0]['tx_cabagfileexplorer_redirect_to'])) === '') {
			return $domain;
		}

		$domain = 'http' . (t3lib_div::getIndpEnv('TYPO3_SSL') ? 's' : '') . '://' . $rowArray[0]['domainName'] . '/' .
			ltrim($rowArray[0]['tx_cabagfileexplorer_redirect_to'], '/');

		return $domain;
	}

	/**
	 * @param integer $pageId
	 *
	 * @return string
	 */
	protected function getRedirectUrl($pageId) {
		return rawurlencode(t3lib_BEfunc::getViewDomain($pageId) . '/index.php?id=' . $pageId);
	}
}

if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/cabag_loginas/class.tx_cabagloginas.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/cabag_loginas/class.tx_cabagloginas.php']);
}
